### PR TITLE
implementing a revert operation 

### LINKED ISSUE #826: CLI helper to revert operations from a particular timestamp
We want to be able to rollback some changes that were made since a particular timestamp.

Since clients expect that timestamp never go back in the past, the idea would be to apply the reverse operations.

It could be a command that takes the history entries from that particular timestamp and apply the opposite operations (deleted records get recreated, etc.).

`$ kinto revert --auth user:pass --bucket=test --collection=images 123456789`

The command would fail if the history is not enabled or if the oldest timestamp of the history is superior that the one specified...

Thoughts ?

### PR DESCRIPTION
**Pull request by oronsan**
Fixes #826 
- [ ] Add documentation.
- [ ] Add tests.
- [ ] Add a changelog entry.
- [ ] Add your name in the contributors file.
- [ ] If you changed the HTTP API, update the [API_VERSION](https://github.com/Kinto/kinto/blob/master/kinto/__init__.py#L15) constant and add an API changelog entry [in the docs](https://github.com/Kinto/kinto/blob/master/docs/api/index.rst)

r? @

### CODE REVIEW

**Comment by oronsan** [comment]
I am currently working on the revert Endpoint in the rest api.
I am writing the revert_post function in the kinto/plugins/history/views.py file.

I want to use the existing history endpoint with the python requests module.
i.e:
- creating a get request
        requests.get('http://localhost:8888/v1/buckets/default/history', auth=('token', 'my-secret'))
- receiving the json response 
- create a list of operations to revert the changes
- apply the list of changes in a batch operation

my question is:
is it appropriate to use the python requests module for this?
or is there another way that is preferable or better than that?

thank you :)
Oron

File: kinto/__main__.py [hunk 1]
```diff
@@ -141,6 +157,14 @@ def main(args=None):
                                          parsed_args['bucket'],
                                          parsed_args['collection'])
 
+    elif which_command == 'revert':
+        env = bootstrap(config_file)
+        return scripts.revert(env,
+                              auth=parsed_args['auth'],
+                              bucket=parsed_args['bucket'],
+                              collection=parsed_args['collection'],
+                              since=parsed_args['since'])
```
Comment (leplatrem) [inline kinto/__main__.py]: If we follow the *URL endpoint* approach (which is better indeed), then the command-line is probably useless.

File: kinto/plugins/history/__init__.py [hunk 2]
```diff
@@ -9,6 +9,11 @@ def includeme(config):
         description='Track changes on data.',
         url='http://kinto.readthedocs.io/en/latest/api/1.x/history.html')
 
+    config.add_api_capability(
+        'revert',
+        description='discard all changes after given timestamp.',
+        url='http://kinto.readthedocs.io/en/latest/api/1.x/revert.html')
+
```
Comment (leplatrem) [inline kinto/plugins/history/__init__.py]: I'm not sure that we need this

File: kinto/plugins/history/views.py [hunk 3]
```diff
@@ -43,3 +45,86 @@ def _extract_filters(self, queryparams=None):
             filters_str_id.append(filt)
 
         return filters_str_id
+
+
+class Revert(object):
+
+    CREATE = 'create'
+    UPDATE = 'update'
+    DELETE = 'delete'
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: These are available in the `ACTIONS` enum in `kinto.core.events`

File: kinto/plugins/history/views.py [hunk 4]
```diff
@@ -43,3 +45,86 @@ def _extract_filters(self, queryparams=None):
...
+                     path='/buckets/{bucket_id}/collections/{collection_id}/revert',
+                     description="Revert a collection to a timestamp, "
+                                 "all later records will be discarded")
+
+def raise_error(request, name, description):
+    error_details = {
+        'name': name,
+        'description': description,
+    }
+    raise_invalid(request, **error_details)
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: I think this does not bring much value

File: kinto/plugins/history/views.py [hunk 5]
```diff
@@ -43,3 +45,86 @@ def _extract_filters(self, queryparams=None):
...
+            print passed_since
+            if passed_since and len(revert_lists['not_paired']) == 0:
+                break
+            self.insert_rec_to_revert_lists(rec, revert_lists, passed_since)
+
+        return revert_lists['paired']
+
+
+revert_service = Service(name="revert",
+                     path='/buckets/{bucket_id}/collections/{collection_id}/revert',
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: I think the revert operation should be at the same level as history. Thus `/buckets/{bucket_id}/revert`.

File: kinto/plugins/history/views.py [hunk 6]
```diff
@@ -43,3 +45,86 @@ def _extract_filters(self, queryparams=None):
...
+    }
+    raise_invalid(request, **error_details)
+
+@revert_service.post()
+def revert_post(request):
+    bucket_id = request.matchdict['bucket_id']
+    collection_id = request.matchdict['collection_id']
+    if 'since' not in request.params:
+        raise_error(request, 'since', 'since request parameter must be provided')
+    else:
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: nitpick: `else` is superfluous since we `raise` otherwise

File: kinto/plugins/history/views.py [hunk 7]
```diff
@@ -48,23 +48,83 @@ def _extract_filters(self, queryparams=None):
...
+        since = int(request.params['since'])
+
+    (host, port) = ('localhost', '8888') #(request.get_hostname(), request.get_port())
+    history = requests.get('http://%s:%s/v1/buckets/default/history?collection_id=%s'
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: Yes you can using the storage backed.

Something like that (*untested*):
```py
storage = request.registry.storage
filters = [Filter('collection_id', collection_id, COMPARISON.EQ)]
entries, total_count = storage.get_all(parent_id='/buckets/%s' % bucket_id,
                                       collection_id='history',
                                       filters=filters)
```

**Comment by leplatrem** [comment]
Do you need help to get started with the tests?

**Comment by oronsan** [comment]
Well, I'm trying to figure out how the tests for the history plugin work.
I'd love it if you have anything that can help me, even a more elaborate explanation 
about the existing tests

Thank you!

**Comment by oronsan** [comment]
also, I'm trying to create a "batch" post request 
in order to make all changes in a batch.

How should I do it using the registry or request object?
I couldn't find  the right way

right now I can't write any tests since my code does not actually do the revert operations...

@leplatrem 

Thank you

**Comment by leplatrem** [comment]
You can start small. I suggest that you make a backup of all this code that you have here, and restart from scratch with very little code.

For example, you start a test that inherits from `HistoryWebTest` in `test_history.py`.

In the `setUp`of this new test, you perform some operations:
* create a bucket
* create a collection
* create 1 record
* delete 1 record
(*see examples in the FilteringTest setup*)

Then you first assertion will be simple: test that you can revert the delete operation.
```py
def test_can_revert_record_deletion(self):
    # assert that the record is missing
    self.app.get(record_url, headers=self.headers, status=404)
    # call the revert endpoint
    # assert that the record is back
    self.app.get(record_url, headers=self.headers, status=200)
```

Once you get that simple bit working, you can iterate like this:
* add a new test that will fail because the code does not handle the new case
* make the test pass with some new code
* refactor and clean-up if necessary
* go to step 1

If our contributions docs do not allow to get started, please let us know and open dedicated issues, even if it's just a question! It is very important that no particular knownledge is necessary to get started!

Good luck and enjoy!

File: kinto/plugins/history/views.py [hunk 8]
```diff
@@ -43,3 +45,88 @@ def _extract_filters(self, queryparams=None):
...
+            if e.content_type == 'application/json':
+                resp = e
+            else:
+                resp = errors.http_error(e)
+        return build_response(resp, subrequest)
+
+    def revert_changes(self):
+        revert_list = self.build_revert_list()
+        if len(revert_list) == 0:
+            raise_invalid(self.request, 'params', 'no history', 'there are no history records to revert after since timestamp')
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: Maybe this is not a problem after all. I didn't see any test for this for example :)

File: kinto/plugins/history/views.py [hunk 9]
```diff
@@ -43,3 +45,88 @@ def _extract_filters(self, queryparams=None):
...
+
+class Revert(resource.UserResource):
+
+    def __init__(self, request):
+        self.request = request
+        self.bucket_id = request.matchdict['bucket_id']
+        if 'since' not in request.params:
+            raise_invalid(self.request, 'params', 'since', 'since request parameter must be provided')
+        self.since = int(request.params['since'])
+        filters = super(Revert, self)._extract_filters(request.params)
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: If you inherit from `UserResource` just to access that private method, then we should refactor that. We could have a `extract_filters` function in `kinto.core.resource` instead, that both places would call

File: kinto/plugins/history/views.py [hunk 10]
```diff
@@ -43,3 +45,88 @@ def _extract_filters(self, queryparams=None):
...
         return filters_str_id
+
+
+class Revert(resource.UserResource):
+
+    def __init__(self, request):
+        self.request = request
+        self.bucket_id = request.matchdict['bucket_id']
+        if 'since' not in request.params:
+            raise_invalid(self.request, 'params', 'since', 'since request parameter must be provided')
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: Maybe we should use `_since` like for filtering resources plural endpoints?

File: kinto/plugins/history/views.py [hunk 11]
```diff
@@ -43,3 +45,88 @@ def _extract_filters(self, queryparams=None):
...
+            if passed_since and len(revert_lists['not_paired']) == 0:
+                break
+            self.insert_rec_to_revert_lists(rec, revert_lists, passed_since)
+
+        return revert_lists['paired']
+
+
+revert_service = Service(name="revert",
+                     path='/buckets/{bucket_id}/revert',
+                     description="Revert a collection to a timestamp, "
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: nitpick: indentation

Comment (leplatrem) [inline kinto/plugins/history/views.py]: the revert is on the bucket, so it would be `Revert a bucket...`

File: kinto/plugins/history/views.py [hunk 12]
```diff
@@ -43,3 +45,88 @@ def _extract_filters(self, queryparams=None):
...
+                                 "all later records will be discarded")
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: nit: objects instead of records, since groups and collections are concerned

File: kinto/plugins/history/views.py [hunk 13]
```diff
@@ -43,3 +45,88 @@ def _extract_filters(self, queryparams=None):
...
+            elif rec['action'] == ACTIONS.CREATE.value:
+                revert_lists['paired'].append(self.create_revert_operation(rec, None, passed_since))
+            else:
+                revert_lists['not_paired'][rec_id] = rec
+
+    def build_revert_list(self):
+        revert_lists = {'not_paired': {}, 'paired': []}
+        passed_since = False
+        for rec in self.history:
+            if not passed_since and rec['last_modified'] < self.since:
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: Can you explain this `passed_since` boolean in comments? I see that you look-up some entries in the history, that you reconciliate later, but this is not easy to grasp

File: kinto/plugins/history/views.py [hunk 14]
```diff
@@ -43,3 +45,88 @@ def _extract_filters(self, queryparams=None):
...
+    def revert_changes(self):
+        revert_list = self.build_revert_list()
+        if len(revert_list) == 0:
+            raise_invalid(self.request, 'params', 'no history', 'there are no history records to revert after since timestamp')
+        return self.batch_revert(revert_list)
+
+    def create_revert_operation(self, last, first, passed_since):
+        path = last['uri']
+        last_is_create = last['action'] == ACTIONS.CREATE.value
+        if last_is_create or (not passed_since and first['action'] == ACTIONS.CREATE.value):
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: please store the second part of this condition to help reading and give some meaning

File: kinto/plugins/history/views.py [hunk 15]
```diff
@@ -43,3 +45,88 @@ def _extract_filters(self, queryparams=None):
...
+                resp = errors.http_error(e)
+        return build_response(resp, subrequest)
+
+    def revert_changes(self):
+        revert_list = self.build_revert_list()
+        if len(revert_list) == 0:
+            raise_invalid(self.request, 'params', 'no history', 'there are no history records to revert after since timestamp')
+        return self.batch_revert(revert_list)
+
+    def create_revert_operation(self, last, first, passed_since):
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: nit: operation → request

File: kinto/plugins/history/views.py [hunk 16]
```diff
@@ -43,3 +45,88 @@ def _extract_filters(self, queryparams=None):
...
+        filters = super(Revert, self)._extract_filters(request.params)
+        filters = [s for s in filters
+                                 if s.field != 'since']
+        self.storage = request.registry.storage
+        self.history, _ = self.storage.get_all(parent_id='/buckets/%s' % self.bucket_id,
+                                               collection_id='history', filters=filters,
+                                               sorting=[Sort('last_modified', -1)])
+
+    def batch_revert(self, revert_list):
+        batch = {'path': '/batch', 'method': 'POST', 'body': {'requests': revert_list}}
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: I think the code could be greatly simplified if this method would be in charge of building the requests from a list of data structures like tuples or dicts. This would prevent to have those `self.create_revert_operation()` everywhere

File: kinto/plugins/history/views.py [hunk 17]
```diff
@@ -43,3 +45,88 @@ def _extract_filters(self, queryparams=None):
...
+
+    def build_revert_list(self):
+        revert_lists = {'not_paired': {}, 'paired': []}
+        passed_since = False
+        for rec in self.history:
+            if not passed_since and rec['last_modified'] < self.since:
+                passed_since = True
+            if passed_since and len(revert_lists['not_paired']) == 0:
+                break
+            self.insert_rec_to_revert_lists(rec, revert_lists, passed_since)
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: The code would be simpler if the `if passed_since` that you have in `insert_rec_to_revert_lists` would be put here

File: kinto/plugins/history/views.py [hunk 18]
```diff
@@ -43,3 +45,88 @@ def _extract_filters(self, queryparams=None):
...
+            revert_lists['paired'].append(self.create_revert_operation(last_change, rec, passed_since))
+        else:
+            if rec_id in revert_lists['not_paired']:
+                if rec['action'] == ACTIONS.CREATE.value:
+                    last_change = revert_lists['not_paired'].pop(rec_id)
+                    revert_lists['paired'].append(self.create_revert_operation(last_change, rec, passed_since))
+            elif rec['action'] == ACTIONS.CREATE.value:
+                revert_lists['paired'].append(self.create_revert_operation(rec, None, passed_since))
+            else:
+                revert_lists['not_paired'][rec_id] = rec
```
Comment (leplatrem) [inline kinto/plugins/history/views.py]: This part deserves some comments, it's quite hard to follow :)

File: tests/plugins/test_history.py [hunk 19]
```diff
@@ -667,3 +667,100 @@ def test_history_entries_are_not_listed_in_permissions_endpoint(self):
...
+        self.rec_body = {'data': {'foo': 42}}
+        self.record = self.app.put_json(self.record_uri, self.rec_body, headers=self.headers).json
+
+        self.revert_uri = '%s/revert' % self.bucket_uri
+
+    def compare_without_last_modified(self, rec1, rec2):
+        last_mod1 = rec1['data'].pop('last_modified')
+        last_mod2 = rec2['data'].pop('last_modified')
+        if 'read' not in rec1['permissions']:
+            rec1['permissions']['read'] = []
```
Comment (leplatrem) [inline tests/plugins/test_history.py]: you can use `rec1['permissions'].setdefault('read', [])`

File: tests/plugins/test_history.py [hunk 20]
```diff
@@ -667,3 +667,100 @@ def test_history_entries_are_not_listed_in_permissions_endpoint(self):
...
+        revert_ops = resp.json['body']['responses']
+        assert len(revert_ops) == 1
+        self.compare_without_last_modified(revert_ops[0]['body'], self.record)
+        all_records = self.app.get(self.collection_uri + '/records', headers=self.headers).json
+        assert len(all_records['data']) == 1
+        self.app.get(self.record_uri, headers=self.headers, status=200)
+
+    def check_if_deleted(self, deleted_rec, revert_response_rec):
+        assert revert_response_rec['body']['data']['deleted'] == True
+        assert revert_response_rec['body']['data']['id'] == deleted_rec['data']['id']
```
Comment (leplatrem) [inline tests/plugins/test_history.py]: if this is used only once, then maybe it can be inlined in the test directly

File: tests/plugins/test_history.py [hunk 21]
```diff
@@ -667,3 +667,100 @@ def test_history_entries_are_not_listed_in_permissions_endpoint(self):
...
+        revert_uri = self.revert_uri + '?since=%s' % ts
+        resp = self.app.post(revert_uri, headers=self.headers)
+        revert_ops = resp.json['body']['responses']
+        assert len(revert_ops) == 1
+        self.check_if_deleted(test_record, revert_ops[0])
+        all_records = self.app.get(self.collection_uri + '/records', headers=self.headers).json
+        assert len(all_records['data']) == 1
+        self.app.get(test_record_uri, headers=self.headers, status=404)
+
+    def compare_body_only(self, rec1, rec2):
```
Comment (leplatrem) [inline tests/plugins/test_history.py]: `self` is not used → function?

**Comment by leplatrem** [comment]
Closing for lack of response. Feel free to reopen, we'd be happy to see all this resumed!
